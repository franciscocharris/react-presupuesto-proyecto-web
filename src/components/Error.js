import React from 'react'
import PropTypes from 'prop-types';

// cuando se va a retornar una sola cosa
export const Error = ({message}) => (
    <p className='alert alert-danger error'>{message}</p>
)

Error.propTypes = {
    message: PropTypes.string.isRequired
}