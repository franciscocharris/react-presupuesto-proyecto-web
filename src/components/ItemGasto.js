import React from 'react'
import PropTypes from 'prop-types';

export const ItemGasto = ({gasto:{nombre = '', cantidad = ''}}) => {
    return (
        <li className='gastos'>
            <p>
                {nombre}
                <span className='gasto'>${cantidad}</span>
            </p>
        </li>
    )
}

ItemGasto.propTypes = {
    gasto: PropTypes.object.isRequired
}