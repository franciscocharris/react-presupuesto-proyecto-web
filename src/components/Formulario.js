import React, { useState } from 'react'
import PropTypes from 'prop-types';
import  shortid from "shortid";
import { useForm } from '../hooks/useForm'
import { Error } from './Error';

export const Formulario = ({agregarNuevoGasto}) => {

    const [error, setError] = useState(false);

    const [gasto, handleInputChange, reset] = useForm({
        nombre: '',
        cantidad: ''
    });

    const {nombre, cantidad} = gasto;

    const agregarGasto = e => {
        e.preventDefault();

        // validar
        if(cantidad < 1 || isNaN(cantidad) || nombre.trim() === ''){
            setError(true);
            return;
        }else{
            setError(false);
        }
        // construir el gasto
        const gasto = {
            nombre,
            cantidad,
            id: shortid.generate()
        }

        // pasar el gasto al componente principal
        agregarNuevoGasto(gasto);

        // resetear form
        reset();
    }

    return (
        <form
            onSubmit={agregarGasto}
        >
            <h2>Agrega tus gastos aqui</h2>
            
            {error && <Error  message='Ambos campos son obligatorios o cantidad del gasto incorrecto' />}

            <div className='campo'>
                <label>Nombre del Gasto</label>
                <input
                    type='text'
                    className='u-full-width'
                    placeholder='ej: Transporte'
                    name='nombre'
                    value={nombre}
                    onChange={handleInputChange}
                />
            </div>

            <div className='campo'>
                <label>Cantidad del Gasto</label>
                <input
                    type='number'
                    className='u-full-width'
                    placeholder='ej: 270'
                    name='cantidad'
                    value={cantidad}
                    onChange={handleInputChange}
                />
            </div>

            <input
                type='submit'
                className='button-primary u-full-width'
                value='Agregar Gasto'
            />
        </form>
    )
}

Formulario.propTypes = {
    agregarNuevoGasto: PropTypes.func.isRequired
}