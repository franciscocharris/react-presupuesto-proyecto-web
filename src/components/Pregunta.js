import React, { useState } from 'react'
import PropTypes from 'prop-types';
import { Error } from './Error';

export const Pregunta = ({setPresupuesto, setRestante, setShowPregunta}) => {

    // definir state - la cantidad
    const [amount, setAmount] = useState(0);
    const [error, setError] = useState(false); 

    // submit para definir el presupuesto
    const agregarPresupuesto = e => {
        e.preventDefault();

        // validar
        if(amount < 1 || isNaN(amount)){
            setError(true);
            // return para que detenga la ejecuciuon
            return;
        }else{
            setError(false);
        }

        setPresupuesto(amount);
        setRestante(amount);

        // una ves todo este validado
        setShowPregunta(false);
    }

    return (
        <>
            <h2>Coloca tu presupuesto</h2>

            {error && <Error message='El Presupuesto es Incorrecto' />}

            <form
                onSubmit={agregarPresupuesto}
            >
                <input 
                    type='number'
                    className='u-full-width'
                    placeholder='coloca tu presupuesto'
                    onChange={e => setAmount(parseInt(e.target.value))}
                />

                <input
                    type='submit'
                    className='button-primary u-full-width'
                    value='Definir Presupuesto'
                />
            </form>
        </>
    )
}

Pregunta.propTypes = {
    setPresupuesto: PropTypes.func.isRequired,
    setRestante: PropTypes.func.isRequired,
    setShowPregunta: PropTypes.func.isRequired
}