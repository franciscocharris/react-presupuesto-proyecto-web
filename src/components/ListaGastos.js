import React from 'react'
import { ItemGasto } from './ItemGasto'
import PropTypes from 'prop-types';

export const ListaGastos = ({gastos}) => (
    <div className='gastos-realizados'>
        <h2>Listado</h2>
        {
            // debe ser con () para que sea explicito el return, no con {}
            gastos.map(gasto => (
                <ItemGasto
                    key={gasto.id}
                    gasto={gasto}
                />
            ))
        }
    </div>
)

ListaGastos.propTypes = {
    gastos: PropTypes.array.isRequired
}
