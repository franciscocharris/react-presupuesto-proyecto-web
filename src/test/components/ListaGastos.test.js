import React from 'react';
import { shallow } from "enzyme"
import { ListaGastos } from "../../components/ListaGastos"

describe('pruebas en <ListasGastos/>', () => {
    
    test('debe de renderrizar bien', () => {
        const gastos = [ 
            {
                nombre: 'pp',
                cantidad: 100,
                id: 1
            },
             {
                nombre: 'papa',
                cantidad: '120202',
                id: 2
            }
        ],
         wrapper = shallow(<ListaGastos gastos={gastos} />);

         expect(wrapper).toMatchSnapshot();
    });

    test('la longitud del array debe ser igual con el n. de itemgasto', () => {
        const gastos = [ 
            {
                nombre: 'pp',
                cantidad: 100,
                id: 1
            },
             {
                nombre: 'papa',
                cantidad: '120202',
                id: 2
            }
        ],
         wrapper = shallow(<ListaGastos gastos={gastos} />);

         expect(wrapper.find('ItemGasto').length).toBe(gastos.length);
    })
    
    
})
