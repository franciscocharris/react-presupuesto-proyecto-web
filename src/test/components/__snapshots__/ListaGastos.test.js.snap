// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <ListasGastos/> debe de renderrizar bien 1`] = `
<div
  className="gastos-realizados"
>
  <h2>
    Listado
  </h2>
  <ItemGasto
    gasto={
      Object {
        "cantidad": 100,
        "id": 1,
        "nombre": "pp",
      }
    }
    key="1"
  />
  <ItemGasto
    gasto={
      Object {
        "cantidad": "120202",
        "id": 2,
        "nombre": "papa",
      }
    }
    key="2"
  />
</div>
`;
